src/jni_core/src/opendal_schema_history.rs

```diff
@@ -0,0 +1,183 @@
+// Copyright 2025 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::borrow::Cow;
+use std::sync::{Arc, OnceLock};
+
+use bytes::Bytes;
+use jni::objects::{JByteArray, JObject, JString};
+use risingwave_common::config::ObjectStoreConfig;
+use risingwave_common::{DATA_DIRECTORY, STATE_STORE_URL};
+use risingwave_object_store::object::object_metrics::ObjectStoreMetrics;
+use risingwave_object_store::object::{ObjectStoreImpl, build_remote_object_store};
+
+use crate::{EnvParam, JAVA_BINDING_ASYNC_RUNTIME, execute_and_catch, to_guarded_slice};
+
+static OBJECT_STORE_INSTANCE: OnceLock<Arc<ObjectStoreImpl>> = OnceLock::new();
+
+// schema history is internal state, all data is stored under the DATA_DIRECTORY directory.
+fn prepend_data_directory(path: &str) -> String {
+    let data_dir = DATA_DIRECTORY.get().map(|s| s.as_str()).unwrap_or("");
```

Comment:
default to `""` looks very dangerous because buckets are shared in cloud environment, which can cause data to be overwritten. We shoulld add a safe guard to make sure data dir is set instead of using default.